tidb-server/server/conn.go

```diff
@@ -0,0 +1,469 @@
+	//charset, utf-8 default
+	data = append(data, uint8(mysql.DefaultCollationID))
+	//status
+	data = append(data, dumpUint16(mysql.ServerStatusAutocommit)...)
+	//below 13 byte may not be used
+	//capability flag upper 2 bytes, using default capability here
+	data = append(data, byte(defaultCapability>>16), byte(defaultCapability>>24))
+	//filter [0x15], for wireshark dump, value is 0x15
+	data = append(data, 0x15)
+	//reserved 10 [00]
+	data = append(data, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+	//auth-plugin-data-part-2
+	data = append(data, cc.salt[8:]...)
+	//filter [00]
+	data = append(data, 0)
+	err := cc.writePacket(data)
+	if err != nil {
+		return err
+	}
+	return cc.flush()
+}
+
+func (cc *clientConn) readPacket() ([]byte, error) {
+	return cc.pkg.readPacket()
+}
+
+func (cc *clientConn) writePacket(data []byte) error {
+	return cc.pkg.writePacket(data)
+}
+
+func calcPassword(scramble, password []byte) []byte {
+	if len(password) == 0 {
+		return nil
+	}
+
+	// stage1Hash = SHA1(password)
+	crypt := sha1.New()
+	crypt.Write(password)
+	stage1 := crypt.Sum(nil)
+
+	// scrambleHash = SHA1(scramble + SHA1(stage1Hash))
+	// inner Hash
+	crypt.Reset()
+	crypt.Write(stage1)
+	hash := crypt.Sum(nil)
+
+	// outer Hash
+	crypt.Reset()
+	crypt.Write(scramble)
+	crypt.Write(hash)
+	scramble = crypt.Sum(nil)
+
+	// token = scrambleHash XOR stage1Hash
+	for i := range scramble {
+		scramble[i] ^= stage1[i]
+	}
+	return scramble
+}
+
+func (cc *clientConn) readHandshakeResponse() error {
+	data, err := cc.readPacket()
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	pos := 0
+	//capability
+	cc.capability = binary.LittleEndian.Uint32(data[:4])
+	pos += 4
+	//skip max packet size
+	pos += 4
+	//charset, skip, if you want to use another charset, use set names
+	cc.collation = data[pos]
+	pos++
+	//skip reserved 23[00]
+	pos += 23
+	//user name
+	cc.user = string(data[pos : pos+bytes.IndexByte(data[pos:], 0)])
+	pos += len(cc.user) + 1
+	//auth length and auth
+	authLen := int(data[pos])
+	pos++
+	auth := data[pos : pos+authLen]
+	checkAuth := calcPassword(cc.salt, []byte(cc.server.cfgGetPwd(cc.user)))
+	if !bytes.Equal(auth, checkAuth) && !cc.server.skipAuth() {
+		return errors.Trace(mysql.NewDefaultError(mysql.ErAccessDeniedError, cc.conn.RemoteAddr().String(), cc.user, "Yes"))
+	}
+
+	pos += authLen
+	if cc.capability|mysql.ClientConnectWithDB > 0 {
+		if len(data[pos:]) == 0 {
+			return nil
+		}
+
+		cc.dbname = string(data[pos : pos+bytes.IndexByte(data[pos:], 0)])
+	}
+
+	return nil
+}
+
+func (cc *clientConn) Run() {
+	defer func() {
+		r := recover()
+		if r != nil {
+			const size = 4096
+			buf := make([]byte, size)
+			buf = buf[:runtime.Stack(buf, false)]
+			log.Errorf("lastCmd %s, %v, %s", cc.lastCmd, r, buf)
+		}
+		cc.Close()
+	}()
+
+	for {
+		cc.alloc.Reset()
+		data, err := cc.readPacket()
+		if err != nil {
+			if errors2.ErrorNotEqual(err, io.EOF) {
+				log.Info(err)
```

Comment:
There is another type of error that is not io.EOF and should not be logged as error.
I'll add TODO comment.